nonmoodlefeedback.php

```diff
@@ -37,6 +37,10 @@
 $userid = optional_param('userid2', 0, PARAM_INT);
 $instance = optional_param('instance', 0, PARAM_INT);
 
+$usercontext = context_user::instance($userid);
+if (!has_capability('moodle/user:viewdetails', $usercontext)) {
+    throw new moodle_exception('nopermissions', 'error', '', get_string('viewuserdetails', 'role'));
+}
```

Comment:
1. This is re-using existing capability, moodle/user:viewdetails, and that has 'captype' => 'read' but we're checking for a "write" action (saving data to the database). Also, this capability has an existing meaning ([This allows a user to view another user's profile page](https://docs.moodle.org/404/en/Capabilities/moodle/user:viewdetails)). I think this should [add a specific capability](https://moodledev.io/docs/4.4/apis/commonfiles#dbaccessphp) for this, but presumably we only want teachers to be able to save feedback for students in that teacher's courses, so we'd want to check a) the teacher is in the course and has this capability, and b) the student is also in the course. Do we have enough information to check this? Let me know if you need any help.
2. `get_string('viewuserdetails', 'role')` -- this string doesn't exist, making the error message _Sorry, but you do not currently have permissions to do that ([[viewuserdetails]])._
3. A `required_capability_exception` exception is probably more appropriate here. But this check could presumably be done with just `require_capability('moodle/user:viewdetails', context_user::instance($userid))`?

Reply:
Actually, I think the current implementation -- checking `has_capability('moodle/user:viewdetails', $usercontext)` -- prevents any teacher from leaving feedback for their students.

Reply:
Changed approach, let me know what you think